File: Leetcode 0204 Count Primes.py
'''
07/12/2023
'''

"""
n = 10
dp[2]
[False, False, True, True, False, True, False, True, False, True]
dp[3]
[False, False, True, True, False, True, False, True, False, False]
"""


class Solution:
    def countPrimes(self, n: int) -> int:

        if n < 2:
            return 0
        dp = [True] * n
        dp[0] = dp[1] = False
        
        for i in range(2, int(n**0.5) +1):
            if dp[i]:
                dp[i*i:n:i] = [False] * len(dp[i*i:n:i])
            print(dp)

        return dp.count(True)    

File: Leetcode 0204 Count Primes.go
// 2025/07/02

func countPrimes(n int) int {
    if n <= 1{
        return 0
    }
    
    dp := make([]bool, n)
    for i:= 0; i < n; i++{
        dp[i] = true
    }
    dp[0], dp[1] = false, false
    for i:=2; i*i < n; i++{
        if dp[i] == true{
            for j:=i*i; j < n; j+=i{
                dp[j] = false
            }  
        }
    }
    count := 0;
    for _, prime := range dp{
        if prime{
            count += 1
        }
    }
    return count

}
